### DISCUSSION TITLE
Can buffers created from one adapter be used in another adapter (same device)

### DISCUSSION THREAD

**Discussion by BLuchterhand** [gfx-rs/wgpu#3999] [Q&A] [answered]
Hey all! I'm currently running two loops:
1. a headless one for running compute logic and generating data
2. the other for updating state and drawing to the screen. 

Each loop gets its own adapter (mostly because I'm new to both Rust and WGPU and sharing resources hurts my head). The buffers generated for the first adapter are throwing an `Attempted to use invalid buffer` error. I should note that everything worked before splitting adapters and loops, so I know the buffers and bind groups/layouts are all properly set. 

Is this because I'm attempting to migrate buffers between adapters or could there be some other issue at hand?

**Comment by kpreid** [comment]
Every buffer and other resource belongs to a specific `Device`, and you'll get an error if you try to use one with a different device. When you say “another adapter (same device)”, what exactly do you mean? You have to specify one `Adapter` when you create a `Device`.

Your sharing problem may be solvable by using an `Arc<wgpu::Device>` and `Arc<wgpu::Queue>`.

**Reply by BLuchterhand** [reply]
Thank you so much for the reply!

When I say 2 different adapters, I'm referring to as if I created 2 adapters (and accompanying devices/queues) like this:
```
    let adapter = instance
        .request_adapter(&wgpu::RequestAdapterOptions {
            power_preference: wgpu::PowerPreference::default(),
            compatible_surface: None, // This would be set for the second adapter to draw to the surface
            force_fallback_adapter: false,
        })
        .await
        .unwrap();

    let (device, queue) = adapter.request_device(
        &wgpu::DeviceDescriptor::default(),
        None,
    ).await.unwrap();
```

I'll give `Arc<wgpu::Device>` and `Arc<wgpu::Queue>` a look! Thank you very much!

**Reply by kpreid** [reply, marked as answer]
That's not the same device in any way, then. So, yes, you will have to create just one device on one adapter and share it between your two threads/tasks.
